module-3/lab-sklearn-and-unsupervised-learning/your-code/main.ipynb

```diff
@@ -318,7 +318,7 @@
     {
```

Comment:
- Good one 👍 "What is the average ash content for each cluster?". Another way of doing it:
```Python
#your code 
for e in set(lab):
    i=np.where(lab==e)
    print(np.average(wine['data'][i,2]))
#other way 
(np.average(wine['data'][:,2][np.where(kmeans.labels_ == 0)]), np.average(wine['data'][:,2][np.where(kmeans.labels_ == 1)]), 
np.average(wine['data'][:,2][np.where(kmeans.labels_ == 2)]), np.average(wine['data'][:,2][np.where(kmeans.labels_ == 3)]))
```
- "For the prescribed_medicines column, fill the missing values with the value no prescription. For the diagnosis column, fill the missing values with no diagnosis." better in one line when possible and the code is readable enough:
```Python
patients.fillna({'prescribed_medicines': 'no prescription', 'diagnosis': 'no diagnosis'}, inplace=True)
```
- Good to use the label encoder. Other way of doing it in less lines:
```Python
from sklearn.preprocessing import LabelEncoder
label = LabelEncoder()
label.fit([False, True])
patients[['patient_diabetic', 'patient_allergic', 'patient_show', 'is_regular_visit']] = patients[['patient_diabetic', 'patient_allergic', 'patient_show', 'is_regular_visit']].apply(label.fit_transform)
patients.head()
```
- In this one, "In the next cells, do the following:
Create a new column called diagnosis_int that has 0 and 1 based on the values in diagnosis.
Create a new column called patient_gender_int that has 0 and 1 based on the values in patient_gender.
Drop the following columns: doctor_name, diagnosis, prescribed_medicines, and patient_gender." you only need this, no lambda needed at all:
```Python
patients['diagnosis_int'] = np.where(patients.diagnosis == "no diagnosis", 0, 1)
patients['patient_gender_int'] = np.where(patients.patient_gender == "Male", 0, 1)
patients.drop(columns=['doctor_name', 'diagnosis', 'prescribed_medicines', 'patient_gender'], inplace=True)
```
- "Bonus Challenge: Visualize K-Means Clusters" , another way of doing this plot:
fig = plt.figure()
ax = fig.add_subplot(111)
scatter = ax.scatter(patients['patient_weight_kg'], patients['patient_height_sm'],
                     c=patients['labels'])
plt.colorbar(scatter)
```
- We can divide (as a taxonomy) the clustering algorithms in two classes:
1. Clustering based in distance: observations can be represented as points in a dimensional space where the dimensions are the features we have. The distance metric can be (euclidean, manhattan, mahalanobis) to separate the different clusters. K-means belongs to this class and it's the most famous one.
2. Clustering based on density: this algorithms don't into account only the distance among "points" but also detecting the areas where the density of points is higher.
- Find below a link to the main clustering algorithms:
https://towardsdatascience.com/the-5-clustering-algorithms-data-scientists-need-to-know-a36d136ef68
- Sklearn has all this clustering algorithms:
https://scikit-learn.org/stable/modules/classes.html#module-sklearn.cluster
- Try the 3D when you have some free time, the only problem with 3D plots is that human brain doesn't work properly in 3D.